src/main/java/org/opensearch/knn/index/codec/util/KNNVectorAsCollectionOfHalfFloatsSerializer.java

```diff
@@ -0,0 +1,130 @@
+     * @param dimension number of floats to serialize
+     */
+    public void floatToByteArray(float[] input, byte[] output, int dimension) {
+        if (input == null || output == null) {
+            throw new IllegalArgumentException("Input/output buffers cannot be null.");
+        }
+        if (dimension > input.length) {
+            throw new IllegalArgumentException("Count exceeds input float array length.");
+        }
+        if (output.length != input.length * BYTES_IN_HALF_FLOAT) {
+            throw new IllegalArgumentException("Output buffer size mismatch. Must be 2x input length.");
+        }
+
+        if (SIMDEncoding.isSIMDSupported() && SIMDEncoding.convertFP32ToFP16(input, output, dimension)) {
+            return;
+        }
+
+        log.warn("SIMD FP32 to FP16 encoding not supported on this platform, falling back to Java implementation.");
+        floatToByteArrayFallback(input, output, dimension);
+    }
+
+    /**
+     * Converts byte[] to float[] using SIMD optimization if supported, otherwise fallback to Java.
+     * @param input the byte[] containing half-precision encoded data
+     * @param output a float[] containing the decoded float32 values
+     * @param dimension number of floats to deserialize
+     * @param offset offset in the output array where deserialization should start
+     */
+    public void byteToFloatArray(byte[] input, float[] output, int dimension, int offset) {
+        if (input == null || output == null) {
+            throw new IllegalArgumentException("Input/output buffers cannot be null.");
+        }
+        if (input.length % BYTES_IN_HALF_FLOAT != 0) {
+            throw new IllegalArgumentException(
+                "Invalid byte array length ("
+                    + input.length
+                    + "). Must be a multiple of "
+                    + BYTES_IN_HALF_FLOAT
+                    + " to represent float16 values."
+            );
+        }
+        if (offset < 0 || offset + dimension * BYTES_IN_HALF_FLOAT > input.length) {
+            throw new IllegalArgumentException("Offset and dimension exceed input length.");
+        }
+
+        if (SIMDDecoding.isSIMDSupported() && SIMDDecoding.convertFP16ToFP32(input, output, dimension, offset)) {
+            return;
+        }
+
+        log.warn("SIMD FP16 to FP32 decoding not supported on this platform, falling back to Java implementation.");
+        byteToFloatArrayFallback(input, output, dimension, offset);
+    }
+
+    /**
+     * Fallback Java implementation of float[] to byte[] conversion.
+     * Handles both little-endian and big-endian platforms.
+     *
+     * @param input  float array to convert
+     * @param output output byte array to hold FP16-encoded values
+     * @param dimension number of elements to convert
+     */
+    private void floatToByteArrayFallback(float[] input, byte[] output, int dimension) {
+        if (IS_LITTLE_ENDIAN) {
```

Comment:
why we need little endian check here? since we are writing these files as new so we know we will always write either as little endian or big endian.